src/sonic-eventd/src/eventd.cpp

```diff
@@ -65,6 +95,169 @@ eventd_proxy::run()
+        {
+            SWSS_LOG_ERROR("Unable to get DB Connector, e=(%s)\n", e.what());
+        }
+        RET_ON_ERR(m_counters_db != NULL, "Failed to get COUNTERS_DB");
+
+        m_stats_table = make_shared<swss::Table>(
+                m_counters_db.get(), COUNTERS_EVENTS_TABLE);
+        RET_ON_ERR(m_stats_table != NULL, "Failed to get events table");
+
+        m_thr_writer = thread(&stats_collector::run_writer, this);
+    }
+    m_thr_collector = thread(&stats_collector::run_collector, this);
+    rc = 0;
+out:
+    return rc;
+}
+
+void
+stats_collector::run_writer()
+{
+    while (true) {
+        if (m_updated.exchange(false)) {
+            /* Update if there had been any update */
+
+            for (int i = 0; i < COUNTERS_EVENTS_TOTAL; ++i) {
+                vector<FieldValueTuple> fv;
+
+                fv.emplace_back(EVENTS_STATS_FIELD_NAME, to_string(m_lst_counters[i]));
+
+                m_stats_table->set(counter_keys[i], fv);
+            }
+        }
+        if (m_shutdown) {
+            break;
+        }
+        this_thread::sleep_for(chrono::milliseconds(10));
+        /*
+         * After sleep always do an update if needed before checking
+         * shutdown flag, as any counters collected during sleep
+         * needs to be updated.
+         */
+    }
+
+    m_stats_table.reset();
+    m_counters_db.reset();
+}
+
+void
+stats_collector::run_collector()
+{
+    int hb_cntr = 0;
+    string hb_key = string(EVENTD_PUBLISHER_SOURCE) + ":" + EVENTD_HEARTBEAT_TAG;
+    event_handle_t pub_handle = NULL;
+    event_handle_t subs_handle = NULL;
+
+    /*
+     * A subscriber is required to set a subscription. Else all published
+     * events will be dropped at the point of publishing itself.
+     */
+    pub_handle = events_init_publisher(EVENTD_PUBLISHER_SOURCE);
+    RET_ON_ERR(pub_handle != NULL,
+            "failed to create publisher handle for heartbeats");
+
+    subs_handle = events_init_subscriber(false, STATS_COLLECTOR_TIMEOUT);
+    RET_ON_ERR(subs_handle != NULL, "failed to subscribe to all");
+
+    /*
+     * Though we can count off of capture socket, then we need to duplicate
+     * code in event_receive which has the logic to count all missed per
+     * runtime id. It also has logic to retire closed runtime IDs.
+     *
+     * So use regular subscriber API w/o cache but timeout to enable
+     * exit, upon shutdown.
+     */
+    /*
+     * The collector service runs until shutdown.
+     * The only task is to update total_published & total_missed_internal.
+     * The write of these counters into redis is done by another thread.
+     */
+
+    while(!m_shutdown) {
+        event_receive_op_t op;
+        int rc = 0;
+
+        try {
+            rc = event_receive(subs_handle, op);
+        }
+        catch (exception& e)
+        {
+            rc = -1;
+            stringstream ss;
+            ss << e.what();
+            SWSS_LOG_ERROR("Cache save event failed with %s",
```

Comment:
Shouldnt log message should be about event_receive failing

Reply:
Yes. Corrected. Thanks.